plugins/hls-selection-range-plugin/src/Ide/Plugin/SelectionRange.hs

```diff
@@ -0,0 +1,107 @@
+                                                          SelectionRangeParams (..),
+                                                          TextDocumentIdentifier (TextDocumentIdentifier))
+import           Prelude                                 hiding (span)
+
+descriptor :: PluginId -> PluginDescriptor IdeState
+descriptor plId = (defaultPluginDescriptor plId)
+    { pluginHandlers = mkPluginHandler STextDocumentSelectionRange selectionRangeHandler
+    }
+
+selectionRangeHandler :: IdeState -> PluginId -> SelectionRangeParams -> LspM c (Either ResponseError (List SelectionRange))
+selectionRangeHandler ide _ SelectionRangeParams{..} = do
+    let (TextDocumentIdentifier uri) = _textDocument
+    let filePathMaybe = toNormalizedFilePath' <$> uriToFilePath' uri
+    case filePathMaybe of
+        Nothing -> pure . Right . List $ []
+        Just filePath -> liftIO $ do
+            let (List positions) = _positions
+            selectionRanges <- runIdeAction "SelectionRange" (shakeExtras ide) $ getSelectionRanges filePath positions
+            pure . Right . List $ selectionRanges
+
+getSelectionRanges :: NormalizedFilePath -> [Position] -> IdeAction [SelectionRange]
+getSelectionRanges file positions = fmap (fromMaybe []) <$> runMaybeT $ do
+    (HAR{hieAst, refMap}, positionMapping) <- useE GetHieAst file
+    positions' <- MaybeT . pure $ traverse (fromCurrentPosition positionMapping) positions
+    ast <- MaybeT . pure $ getAsts hieAst Map.!? (coerce. mkFastString . fromNormalizedFilePath) file
+    let ast' = runReader (preProcessAST ast) (PreProcessEnv refMap)
+    MaybeT . pure . traverse (toCurrentSelectionRange positionMapping) $
+        findSelectionRangesByPositions (astPathsLeafToRoot ast') positions'
+
+-- | Like 'toCurrentPosition', but works on 'SelectionRange'
+toCurrentSelectionRange :: PositionMapping -> SelectionRange -> Maybe SelectionRange
+toCurrentSelectionRange positionMapping SelectionRange{..} = do
+    newRange <- toCurrentRange positionMapping _range
+    pure $ SelectionRange {
+        _range = newRange,
+        _parent = _parent >>= toCurrentSelectionRange positionMapping
+    }
+
+-- | Build all paths from ast leaf to root
+astPathsLeafToRoot :: HieAST a -> [SelectionRange]
+astPathsLeafToRoot = mapMaybe (spansToSelectionRange . simplifySpans) . go [[]]
+  where
+    go acc (Node _ span [])       = fmap (span:) acc
+    go acc (Node _ span children) = concatMap (go (fmap (span:) acc)) children
+
+spansToSelectionRange :: [Span] -> Maybe SelectionRange
+spansToSelectionRange [] = Nothing
+spansToSelectionRange (span:spans) = Just $
+    SelectionRange {_range = realSrcSpanToRange span, _parent = spansToSelectionRange spans}
+
+simplifySpans :: [Span] -> [Span]
+simplifySpans = foldr go []
+  where
+    go x []     = [x]
+    go x (y:ys) = if x == y then y:ys else x:y:ys
+
+findSelectionRangesByPositions :: [SelectionRange] -> [Position] -> [SelectionRange]
```

Comment:
This could do with a comment, it's not totally obvious what it does: it finds the *first* selection range in the input whose range contains the given position (and then does that for the whole input list). Makes me wonder whether there could be two such selection ranges, does that matter?